### PR TITLE
[mta] Fused SGD

### PR DESCRIPTION
**Pull request by crcrpar**
depends on #116583

rel:
- #94791

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -1289,11 +1289,21 @@ def test_grad_scaling_autocast_fused(self):
     # Compare non-fused optimizer vs fused one as the fused one unscales gradients
     # inside its cuda kernel unlike the other.
     def test_grad_scaling_autocast_fused_optimizers(self):
-        for optimizer_ctor, optimizer_kwargs, separate_unscale in product(
+        for optimizer_ctor, optimizer_kwargs, separate_unscale in list(product(
```
Comment (janeyx99) [inline test/test_cuda.py:1292]: Hmmm this makes me wonder...maybe one day these should live in test/test_optim.py as well...to deduplicate the configuration logic. (This is out of the scope of the PR)

File: torch/testing/_internal/common_optimizers.py
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:1]: Can you also add a test case for tensor lr by adding an OptimizerInput similar to in Adam https://github.com/pytorch/pytorch/blob/b18d8d4595aa6e0768eedd5fc7d4a4402c567181/torch/testing/_internal/common_optimizers.py#L352-L356

**Review by janeyx99 — APPROVED** [review]
Looks great! Thanks for bringing this up to date.

I've left some comments about adding tests, but other than that + the rebase, this should be good to go.

File: test/test_cuda.py [hunk 2]
```diff
@@ -3209,8 +3233,9 @@ def test_graph_scaling_fused_optimizers(self):
                     scaler_for_graphed._lazy_init_scale_growth_tracker(torch.device("cuda"))
 
                 # Control (capturable=False)
-
-                opt = OptClass(params_control, capturable=False, **kwargs)
+                if has_capturable_arg:
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply test/test_cuda.py:3236]: Good point, we kind of do so. 
SGD's `__init__` doesn't have the argument of `capturable` but Adam and AdamW do. From the perspective of optimizer configs, we have two identical SGD optimizers but one is run without graph capturing while the other is run with graph capturing (depending on `actually_do_graphs` value).
It makes sense to get rid of `actually_do_graphs=False` path for SGD.

**Comment by davidmerwin** [comment]
>### Merge started ...

@pytorchmergebot @pytorchmergebot @pytorch-bot 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by crcrpar** [comment]
@janeyx99 could you help me fix the failure on rocm? on my local cuda device, the failing case takes <5seconds.

**Comment by janeyx99** [comment]
Hm, this probably suggests there is some operation either deadlocking or taking a very long time for ROCm. The test on trunk based on logs is only 0.228s previously. @jeffdaily have you seen something like this before?

Also now that I looked at the logs, could you update test_grad_scaling_autocast_fused to contain SGD now too?

**Comment by jeffdaily** [comment]
The following kernel is hanging.  I was able to build this PR and reproduce.

```
void at::native::(anonymous namespace)::multi_tensor_apply_kernel<at::native::(anonymous namespace)::TensorListMetadata<2>, at::native::(anonymous namespace)::FusedSgdMathFunctor<float, 2>, double, double, float*, double, double, bool, bool, bool, float*, float*>(at::native::(anonymous namespace)::TensorListMetadata<2>, at::native::(anonymous namespace)::FusedSgdMathFunctor<float, 2>, double, double, float*, double, double, bool, bool, bool, float*, float*)
```

**Comment by jeffdaily** [comment]
I don't have an explanation yet, but since the kernel had two potential paths (`use_faster_load_store`), I used a preprocessor #ifdef for rocm to force the other path.  This worked.  So there was something wrong with using the faster load store path.  This is my diff that worked for ROCm

```diff
diff --git a/aten/src/ATen/native/cuda/FusedSgdKernel.cu b/aten/src/ATen/native/cuda/FusedSgdKernel.cu
index a00b47ff79..0c52ee5783 100644
--- a/aten/src/ATen/native/cuda/FusedSgdKernel.cu
+++ b/aten/src/ATen/native/cuda/FusedSgdKernel.cu
@@ -87,6 +87,7 @@ struct FusedSgdMathFunctor {
         init_args<depth>(args, tl, chunk_idx, chunk_size, tensor_loc)};
     n -= chunk_idx * chunk_size;

+#ifndef USE_ROCM
     const auto use_faster_load_store =
         (n % kILP == 0) && (chunk_size % kILP == 0) && all_aligned;
     if (use_faster_load_store) {
@@ -116,7 +117,9 @@ struct FusedSgdMathFunctor {
           load_store(args[2], r_args[2], i_start, 0);
         }
       }
-    } else {
+    } else
+#endif
+    {
       for (auto i_start = 0; i_start < n && i_start < chunk_size;
            i_start += blockDim.x * kILP) {
         load_args<depth>(r_args, args, i_start, chunk_size, n);
```

**Comment by crcrpar** [comment]
Thank you Jeff for the patch. It still feels a little bit counterintuitive as the other fused optimizer kernels have the same logic of `use_faster_load_store`.

**Comment by janeyx99** [comment]
This PR looks good to merge, though it would be great to track the rocm issue with faster load store

**Pull request closed without merging**